File: typescript/src/advertisements/user/infrastructure/persistence/SqliteUserRepository.ts
import { UserRepository } from '../../domain/UserRepository';
import { AdminUser } from '../../domain/AdminUser';
import { MemberUser } from '../../domain/MemberUser';
import { DatabaseConnection } from '../../../../framework/database/DatabaseConnection';
import { UserId } from '../../../shared/domain/value-object/UserId';
import { Email } from '../../../shared/domain/value-object/Email';
import { CivicCenterId } from '../../../shared/domain/value-object/CivicCenterId';
import {Role} from "../../domain/value-object/Role";
import {UserStatus} from "../../domain/value-object/UserStatus";
import {MemberNumber} from "../../domain/value-object/MemberNumber";

export class SqliteUserRepository implements UserRepository {
    constructor(private dbConnection: DatabaseConnection) {}

    async findAdminById(id: UserId): Promise<AdminUser | null> {
        const result = await this.dbConnection.query(`SELECT * FROM users WHERE id = ?`, [id.value()]);
        if (!result || result.length < 1) {
            return null;
        }

        const row = result[0] as any;
        if (row.role === 'admin') {
            return AdminUser.fromDatabase(
                new UserId(row.id),
                new Email(row.email),
                Role.ADMIN,
                new CivicCenterId(row.civic_center_id),
                UserStatus.fromString(row.status)
            );
        }

        return null;
    }

    async findMemberById(id: UserId): Promise<MemberUser | null> {
        const result = await this.dbConnection.query(`SELECT * FROM users WHERE id = ?`, [id.value()]);
        if (!result || result.length < 1) {
            return null;
        }

        const row = result[0] as any;
        if (row.role === 'member') {
            return MemberUser.fromDatabase(
                new UserId(row.id),
                new Email(row.email),
                Role.MEMBER,
                new MemberNumber(row.member_number),
                new CivicCenterId(row.civic_center_id),
                UserStatus.fromString(row.status)
            );
        }

        return null;
    }

    async findAdminOrMemberById(id: UserId): Promise<AdminUser | MemberUser | null> {
        const result = await this.dbConnection.query(`SELECT * FROM users WHERE id = ?`, [id.value()]);
        if (!result || result.length < 1) {
            return null;
        }

        const row = result[0] as any;
        if (row.role === 'admin') {
            return AdminUser.fromDatabase(
                new UserId(row.id),
                new Email(row.email),
                Role.ADMIN,
                new CivicCenterId(row.civic_center_id),
                UserStatus.fromString(row.status)
            );
        }

        if (row.role === 'member') {
            return MemberUser.fromDatabase(
                new UserId(row.id),
                new Email(row.email),
                Role.MEMBER,
                new MemberNumber(row.member_number),
                new CivicCenterId(row.civic_center_id),
                UserStatus.fromString(row.status)
            );
        }

        return null;
    }

    async saveMember(member: MemberUser): Promise<void> {
        if (this.isASignUp(member)) {
            await this.dbConnection.execute(
                `INSERT INTO users (id, email, password, role, member_number, civic_center_id, status) VALUES (?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(id) DO UPDATE SET email = excluded.email, role = excluded.role, member_number = excluded.member_number, civic_center_id = excluded.civic_center_id, status = excluded.status`,
                [
                    member.id().value(),
                    member.email().value(),
                    member.password()?.value(),
                    member.role().value(),
                    member.memberNumber().value(),
                    member.civicCenterId().value(),
                    member.status().value()
                ]
            );
            return;
        }

        await this.dbConnection.execute(
            `INSERT INTO users (id, email, password, role, member_number, civic_center_id) VALUES (?, ?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET email = excluded.email, password = excluded.password, role = excluded.role, member_number = excluded.member_number, civic_center_id = excluded.civic_center_id`,
            [
                member.id().value(),
                member.email().value(),
                member.password()?.value(),
                member.role().value(),
                member.memberNumber().value(),
                member.civicCenterId().value()
            ]
        );
    }

    private isASignUp(member: MemberUser): boolean {
        return member.password() == null;
    }
}

File: kotlin/src/main/kotlin/advertisements/user/infrastructure/persistence/SqliteUserRepository.kt
package advertisements.user.infrastructure.persistence

import advertisements.user.domain.*
import advertisements.shared.value_object.*
import advertisements.user.domain.value_object.MemberNumber
import advertisements.user.domain.value_object.Role
import advertisements.user.domain.value_object.UserStatus
import framework.database.DatabaseConnection

class SqliteUserRepository(private val dbConnection: DatabaseConnection) : UserRepository {

    override fun findAdminById(id: UserId): AdminUser? {
        val result = dbConnection.query("SELECT * FROM users WHERE id = '${id.value()}'")
        if (!result.next()) {
            return null
        }

        return if (result.getString("role") == "admin") {
            AdminUser.fromDatabase(
                UserId(result.getString("id")),
                Email(result.getString("email")),
                Role.ADMIN,
                CivicCenterId.create(result.getString("civic_center_id")),
                UserStatus.fromString(result.getString("status"))
            )
        } else {
            null
        }
    }

    override fun findMemberById(id: UserId): MemberUser? {
        val result = dbConnection.query("SELECT * FROM users WHERE id = '${id.value()}'")
        if (!result.next()) {
            return null
        }

        return if (result.getString("role") == "member") {
            MemberUser.fromDatabase(
                UserId(result.getString("id")),
                Email(result.getString("email")),
                Role.MEMBER,
                MemberNumber(result.getString("member_number")),
                CivicCenterId.create(result.getString("civic_center_id")),
                UserStatus.fromString(result.getString("status"))
            )
        } else {
            null
        }
    }

    override fun findAdminOrMemberById(id: UserId): Any? {
        val result = dbConnection.query("SELECT * FROM users WHERE id = '${id.value()}'")
        if (!result.next()) {
            return null
        }

        return when (result.getString("role")) {
            "admin" -> AdminUser.fromDatabase(
                UserId(result.getString("id")),
                Email(result.getString("email")),
                Role.ADMIN,
                CivicCenterId.create(result.getString("civic_center_id")),
                UserStatus.fromString(result.getString("status"))
            )
            "member" -> MemberUser.fromDatabase(
                UserId(result.getString("id")),
                Email(result.getString("email")),
                Role.MEMBER,
                MemberNumber(result.getString("member_number")),
                CivicCenterId.create(result.getString("civic_center_id")),
                UserStatus.fromString(result.getString("status"))
            )
            else -> null
        }
    }

    override fun saveMember(member: MemberUser) {
        val query = if (isASignUp(member)) {
            """
            INSERT INTO users (id, email, password, role, member_number, civic_center_id, status) 
            VALUES ('${member.id().value()}', '${member.email().value()}', '${member.password()?.value()}', '${member.role().value()}', '${member.memberNumber().value()}', '${member.civicCenterId().value()}', '${member.status().value()}') 
            ON CONFLICT(id) DO UPDATE SET email = '${member.email().value()}', role = '${member.role().value()}', member_number = '${member.memberNumber().value()}', civic_center_id = '${member.civicCenterId().value()}', status = '${member.status().value()}';
            """
        } else {
            """
            INSERT INTO users (id, email, password, role, member_number, civic_center_id) 
            VALUES ('${member.id().value()}', '${member.email().value()}', '${member.password()?.value()}', '${member.role().value()}', '${member.memberNumber().value()}', '${member.civicCenterId().value()}') 
            ON CONFLICT(id) DO UPDATE SET email = '${member.email().value()}', password = '${member.password()?.value()}', role = '${member.role().value()}', member_number = '${member.memberNumber().value()}', civic_center_id = '${member.civicCenterId().value()}';
            """
        }
        dbConnection.execute(query)
    }

    private fun isASignUp(member: MemberUser): Boolean {
        return member.password() == null
    }
}
